### ISSUE TITLE
Is there a way to collect data and visualize from CORE?

### ISSUE THREAD

**Issue by harrycha** [coreemu/core#582]
Hi All,

I've read in the documentation there is a state called "datacollect" which creates log files after a session has stopped. What are in these log files and where would I retrieve them? I've also seen this icon on the core-gui but it doesn't trigger any function. 

![image](https://user-images.githubusercontent.com/33790468/122799557-b8015100-d276-11eb-808b-0222e9c2357e.png)

I'd really like to get data out on throughput between links when I add loss or delay, any advice on how to do this would be most appreciated.

**Comment by bharnden** [comment]
datacollect is a state that can be leveraged for such things, like collecting data, but it is not something that occurs on its own. What you run and how you go about collecting data is open ended and up to you.

Some examples:
* run mgen traffic sessions, then collect and process the logs
* run tcpdump/tshark and then collect and process the output

You could use a hook script during datacollect to run a user defined script at a point during shutdown to collect this type of data, but you can also do this at any other time as well, it's all up to the user driving CORE.

The bottom line is that there is no built in functionality that does this for you, as what is wanted to could vary from person to person, so it is up to the users to define and collect what they want.

**Comment by harrycha** [comment]
Great ideas and things to think about. I'll be generating this traffic outside of CORE because I need some headers and specific protocols to be sent. I'd like to run the tcpdump on one of my virtual node's interfaces and then collect that log file. It sounds like I can create a hook script to somewhat automate this action. Where would that tcpdump for a virtual node reside and do you maybe have a demo script that I could look at?

Just a follow up what does that CORE-GUI button do? As you said there's no inbuilt functionality that creates visualizations but what does the graph button represent?

**Comment by bharnden** [comment]
For a tcpdump the output can be captured to a file anywhere, it depends on what command you run. Some people may dump everything to a directory in /tmp and name the files related to each node as one possibility.

I believe the graph button in the legacy GUI will display something when hovering on the links during runtime, I believe it trys to show a notional bandwidth. But it is just an example visual rather than something concrete to capture for analysis etc.

**Comment by harrycha** [comment]
Instead of a tcpdump log file would it be possible to use the python api to get throughput from a link? I can read the values but I'd like to capture them along with the link parameters.

**Comment by harrycha** [comment]
There is a function called throughputs() within the client.py file. This is kind of what I was looking for. The only issue is it outputs all interfaces, do you know how could I choose one?

`    def throughputs(
        self, session_id: int, handler: Callable[[core_pb2.ThroughputsEvent], None]
    ) -> grpc.Future:
        """
        Listen for throughput events with information for interfaces and bridges.
        :param session_id: session id
        :param handler: handler for every event
        :return: stream processing events, can be used to cancel stream
        :raises grpc.RpcError: when session doesn't exist
        """
        request = core_pb2.ThroughputsRequest(session_id=session_id)
        stream = self.stub.Throughputs(request)
        start_streamer(stream, handler)
        return stream`

Again thanks for all the help!

**Comment by bharnden** [comment]
That is used by the GUI for some rough data, I would personally not depend on that for your data. CORE doesn't try to facilitate this part for you, you could look at the server side code for this, but there may be a better command/way for you to collect your exact desires.
